### ISSUE TITLE
HTTP::Client Post method fails if you call more than once in a single crystal program

### ISSUE THREAD

**Issue by perks** [crystal-lang/crystal#2243]
Running these two calls separately by calling separate files works:
`crystal test1.cr`
```crystal
# This works fine
client = HTTP::Client.new "api.somehost.io", ssl: true

client.before_request do | request |
  request.headers[ "X-CUSTOM-Authorization"] = "SECRET"
end

response = client.post("/gm/instances/23936/restart")
print response
print response.body
````
then
`crystal test2.cr`
```crystal
client = HTTP::Client.new "api.somehost.io", ssl: true

client.before_request do | request |
  request.headers[ "X-CUSTOM-Authorization"] = "SECRET"
end

response = client.post("/gm/instances/23936/stop")
print response
print response.body
````

But running this returns an Index out of bounds error:

`crystal test_combined.cr`
```crystal
client = HTTP::Client.new "api.somehost.io", ssl: true

client.before_request do | request |
  request.headers[ "X-CUSTOM-Authorization"] = "SECRET"
end

response = client.post("/gm/instances/23936/restart")
print response
print response.body
# This succeeds perfectly fine

# Waiting enough time in between
sleep 10

#This fails
response = client.post("/gm/instances/23936/stop")
print response
print response.body
```
Stacktrace:
```
[4376761682] *CallStack::unwind:Array(Pointer(Void)) +82
[4376761585] *CallStack#initialize<CallStack>:Array(Pointer(Void)) +17
[4376761544] *CallStack::new:CallStack +40
[4376771985] *Exception@Exception#initialize<IndexError, String>:CallStack +33
[4376771917] *IndexError#initialize<IndexError, String>:CallStack +29
[4376771852] *IndexError::new<String>:IndexError +92
[4376771740] *IndexError::new:IndexError +28
[4376854699] *Array(String)@Array(T)#at<Array(String), Int32>:String +107
[4376854585] *Array(String)@Array(T)#[]<Array(String), Int32>:String +9
[4376919595] *HTTP::Client::Response::from_io<(OpenSSL::SSL::Socket | TCPSocket), Bool, Bool>:HTTP::Client::Response +219
[4376873142] *HTTP::Client#exec_internal<HTTP::Client, HTTP::Request>:HTTP::Client::Response +150
[4376872742] *HTTP::Client#exec<HTTP::Client, HTTP::Request>:HTTP::Client::Response +38
[4376872205] *HTTP::Client#exec<HTTP::Client, String, String, Nil, Nil>:HTTP::Client::Response +29
[4376872167] *HTTP::Client#post<HTTP::Client, String>:HTTP::Client::Response +39
[4376720829] __crystal_main +45117
[4376729848] main +40
```

**Comment by asterite** [comment]
@perks Does it also happen with GET requests? Does it happen with less time waiting between the calls? Does it only happen with `ssl: true`? I tried to reproduce locally but couldn't. Any https server we can test this against?

My guess is that the socket is being closed on the other end. I tried with GET against google, waiting 2 minutes in between, and I get an error too, although it's different:

```crystal
require "http/client"

client = HTTP::Client.new "www.google.com"

loop do
  puts client.get("/").body # second time give: unexpected end of http response (Exception)
  sleep 2.minutes
end
```

We should probably reconnect on a disconnection, but I'm not sure it's the same problem you have.

**Comment by perks** [comment]
hey @asterite, it happens with no wait between the calls either (I tried this first actually), and unfortunately the server doesn't accept non https requests. If you want, I can PM you temporary a temporary credential key to test the endpoint, along with some of the docs concerning it (i'm writing an online CTF API wrapper)

**Comment by asterite** [comment]
@perks Sure, just send me an email and I'll try to debug it and see what's wrong

**Comment by asterite** [comment]
@perks Thank you!! With your info I was able to reproduce it locally and saw that instead of a regular response status line for the next response I got a "\r\n". I wondered why... and I immediately thought that the problem might be with chunked response. I checked the specs and we were missing reading a final "\r\n". I also was able to reproduce it against a local http server that just streamed chunked content. After the fix, the test against the local server started working, and also the code you gave me.
